### PR TITLE
i#2941: properly recognize DR scratch register slot in drreg

### LINKED ISSUE #2933: drreg state restore problems with DR spill slots
First, drreg fails to properly identify DR spill slots, b/c it assumes the
TLS slots are ordered low to high.

Second, when that bug is fixed, it leads to restoring regs it wasn't before
(b/c it thought they weren't its spills) which leads to an assert:

```
<Application /work/dr/git/build_x64_dbg_tests/clients/bin64/tool.drcacheoff.burst_threads (16044) DynamoRIO usage error : dr_read_saved_reg(): drcontext does not belong to current thread>

#5  0x00000000004b2803 in external_error (file=0x73aac8 "/work/dr/git/src/core/lib/instrument.c", line=5734, 
    msg=0x73c910 "dr_read_saved_reg(): drcontext does not belong to current thread") at /work/dr/git/src/core/utils.c:200
#6  0x00000000005be4f1 in dr_read_saved_reg (drcontext=0x475c2080, slot=SPILL_SLOT_1) at /work/dr/git/src/core/lib/instrument.c:5733
#7  0x00000000006d6580 in get_spilled_value (drcontext=0x475c2080, slot=4) at /work/dr/git/src/ext/drreg/drreg.c:253
#8  0x00000000006db73e in drreg_event_restore_state (drcontext=0x475c2080, restore_memory=0 '\000', info=0x7f3feacec130)
    at /work/dr/git/src/ext/drreg/drreg.c:1767
#9  0x00000000006def9f in drmgr_restore_state_event (drcontext=0x475c2080, restore_memory=0 '\000', info=0x7f3feacec130)
    at /work/dr/git/src/ext/drmgr/drmgr.c:1864
#10 0x00000000005b54dc in instrument_restore_state (dcontext=0x475c2080, restore_memory=false, info=0x7f3feacec130)
    at /work/dr/git/src/core/lib/instrument.c:1715
#11 0x00000000005c91c6 in recreate_app_state_internal (tdcontext=0x475c2080, mcontext=0x7f3feacec800, just_pc=false, owning_f=0x0, 
    restore_memory=false) at /work/dr/git/src/core/translate.c:1208
#12 0x00000000005c9598 in recreate_app_state (tdcontext=0x475c2080, mcontext=0x7f3feacec800, restore_memory=false, f=0x0)
    at /work/dr/git/src/core/translate.c:1331
#13 0x00000000005a2c99 in at_safe_spot (trec=0x474ae128, mc=0x7f3feacec800, desired_state=THREAD_SYNCH_SUSPENDED_VALID_MCONTEXT)
    at /work/dr/git/src/core/synch.c:616
#14 0x00000000005a4156 in synch_with_thread (id=16047, block=false, hold_initexit_lock=true, caller_state=THREAD_SYNCH_NONE, 
    desired_state=THREAD_SYNCH_SUSPENDED_VALID_MCONTEXT, flags=17) at /work/dr/git/src/core/synch.c:1034
#15 0x00000000005a5801 in synch_with_all_threads (desired_synch_state=THREAD_SYNCH_SUSPENDED_VALID_MCONTEXT, 
    threads_out=0x7f3feacece90, num_threads_out=0x7f3feacece8c, cur_state=THREAD_SYNCH_NO_LOCKS_NO_XFER, flags=20)
    at /work/dr/git/src/core/synch.c:1414
#16 0x00000000005a7baf in detach_on_permanent_stack (internal=true, do_cleanup=true) at /work/dr/git/src/core/synch.c:2058
```

That assert should be removed: dcontext for another thread is a valid usage
scenario for synchall.

The underlying issues is that drreg sees what looks like an rax spill but no restore:
```
   0x4ef7373f:  movabs %rax,%gs:0x8
   0x4ef7374a:  lahf   
   0x4ef7374b:  seto   %al
   0x4ef7374e:  cmp    %gs:0x8,%rcx
   0x4ef73757:  jne    0x4ef3830f
```

Here is the source, x64 stay-on-trace cmp:
```
 +267  L3                      65 48 a3 00 00 00 00 mov    %rax -> %gs:0x00[8byte]
                               00 00 00 00
 +278  L3                      48 b8 a4 45 61 4c 60 mov    $0x00007f604c6145a4 -> %rax
                               7f 00 00
 +288  L3                      65 48 a3 08 00 00 00 mov    %rax -> %gs:0x08[8byte]
                               00 00 00 00
 +299  L3                      9f                   lahf    -> %ah
 +300  L3                      0f 90 c0             seto    -> %al
 +303  L3                      65 48 3b 0c 25 08 00 cmp    %rcx %gs:0x08[8byte]
                               00 00
 +312  L4 @0x0000000054505698  0f 85 19 63 f8 ff    jnz    $0x00000000543f8f0f <shared_trace_cmp_ret>
 +318  L3                      65 48 8b 0c 25 10 00 mov    %gs:0x10[8byte] -> %rcx
                               00 00
 +327  L3                      65 48 a1 00 00 00 00 mov    %gs:0x00[8byte] -> %rax
                               00 00 00 00
```

So that's not a spill: DR is just using that TLS slot instead of a 3rd
scratch reg.  That's TLS_REG1_SLOT, which is the 3rd and final one exposed
to clients.

Note that drreg is already handling DR slots not lasting across app instrs
by non-lazily restoring, but it looks like we have fundamental issues with
state restore and DR slots.  For simple DR uses like a local spill and
restore (such as for rip-rel, though that uses a slot not exposed to
clients), a state restore in between would simply be restored twice, once
by DR and once by drreg, with no ill effects.

But, we have this no-restore DR use for trace cmp, and another tool
component could use DR slots too.

Is the only safe thing to do to require enough dedicated drreg-only raw-TLS
slots to avoid ever using DR slots?  Just avoiding the 3rd slot would
reduce the likelihood but not eliminate the theoretical problems.

It would be nice if, at least for recreate and not stored-info state
restores, we could combine extra info during recreation with the raw
mcontext snapshot

### LINKED ISSUE #2941: burst_threadfilter test failing with "Cannot correctly handle a received signal"
Xref #26

https://travis-ci.org/DynamoRIO/dynamorio/jobs/368242799
```
  5/290 Test #250: code_api|tool.drcacheoff.burst_threadfilter ..................***Failed   11.62 sec
Running cmd |/home/travis/build/DynamoRIO/dynamorio/build_debug-internal-64/clients/bin64/tool.drcacheoff.burst_threadfilter|
CMake Error at /home/travis/build/DynamoRIO/dynamorio/suite/tests/runmulti.cmake:106 (message):
  *** cmd failed (40): pre-DR init
  pre-DR start
  pre-DR detach
  pre-DR init
  pre-DR start
  pre-DR detach
  pre-DR init
  pre-DR start
  pre-DR detach
  <ERROR: master_signal_handler with no siginfo (i#26?): tid=18964, sig=11>
  <Application
  /home/travis/build/DynamoRIO/dynamorio/build_debug-internal-64/clients/bin64/tool.drcacheoff.burst_threadfilter
  (18958).  Cannot correctly handle a received signal.>
  ***
```

### PR DESCRIPTION
**Pull request by hgreving2304**
Fix drreg to properly recognize and ignore the 3rd DR slot if a spill is detected but no restore is found. This does not fix #2933 as it is still possible to break drreg with a client that is using DR spill slots.

Add test for above.

Issue: #2933, #511
Fixes #2941

### CODE REVIEW

**Comment by hgreving2304** [comment]
I have moved clang-format awkwardly and uselessly to the last line of drreg-test.c. The reason is I believe a bug in clang-format that cause it to appear a tab difference where there is clearly none. I left the line in the file even though the last line is useless, but to have a complete off/on sequence.

**Comment by derekbruening** [comment]
> I have moved clang-format awkwardly and uselessly to the last line of drreg-test.c. The reason is I believe a bug in clang-format that cause it to appear a tab difference where there is clearly none. I left the line in the file even though the last line is useless, but to have a complete off/on sequence.

Please explain this in a comment in the file (or someone may move it back as it looks like an error, not matching all the other files with no explanation).

File: suite/tests/client-interface/drreg-test.c [hunk 1]
```diff
@@ -334,6 +368,32 @@ GLOBAL_LABEL(FUNCNAME:)
...
+        add      REG_XSP, FRAME_PADDING /* make a legal SEH64 epilog */
+        POP_CALLEE_SAVED_REGS()
+        ret
+#endif
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
 END_FILE
-/* clang-format on */
 #endif
+/* clang-format on */
```
Comment (hgreving2304) [inline suite/tests/client-interface/drreg-test.c]: See D's comment below

Comment (derekbruening) [reply suite/tests/client-interface/drreg-test.c]: A comment in the *source* file.

Also, the on and off no longer match post-preprocessor: we run this file through the preprocessor and the C part now has an on with no preceding off (while the asm part has an off with no on).  We don't run clang-format at that level so it doesn't really matter but it feels wrong: seems nicer to also move the off before the ifdef.

File: suite/tests/client-interface/drreg-test.dll.c [hunk 2]
```diff
@@ -265,6 +265,16 @@ event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst
             res = drreg_unreserve_aflags(drcontext, bb, inst);
             CHECK(res == DRREG_SUCCESS, "unreserve of aflags should work");
         }
+    } else if (subtest == DRREG_TEST_6_C) {
+        /* Save the 3rd DR slot at the label, restore register after
+         * the xl8 point this test.
```
Comment (derekbruening) [inline suite/tests/client-interface/drreg-test.dll.c]: I don't understand "the xl8 point this test": some words are missing?

File: ext/drreg/drreg.c [hunk 3]
```diff
@@ -1609,9 +1609,11 @@ is_our_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT,
                 opnd_get_disp(dr_reg_spill_slot_opnd(drcontext, SPILL_SLOT_1));
             uint DR_max_offs = opnd_get_disp(
                 dr_reg_spill_slot_opnd(drcontext, dr_max_opnd_accessible_spill_slot()));
-            if (DR_min_offs > DR_max_offs)
-                DR_min_offs = DR_max_offs;
-            slot = (offs - DR_min_offs) / sizeof(reg_t);
+            if (DR_min_offs > DR_max_offs) {
```
Comment (derekbruening) [inline ext/drreg/drreg.c]: commit message error: "but not restore" => "but no restore"

File: suite/tests/client-interface/drreg-test.c (hunk 1 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.c]: Done

File: ext/drreg/drreg.c (hunk 3 above)
Comment (hgreving2304) [reply ext/drreg/drreg.c]: Thanks

File: suite/tests/client-interface/drreg-test.dll.c (hunk 2 above)
Comment (hgreving2304) [reply suite/tests/client-interface/drreg-test.dll.c]: "in"

**Pull request merged**